extra/mariadbfrm/mariadb_frm.cc

```diff
@@ -0,0 +1,531 @@
+            filename, strerror(errno));
+    return NULL;
+  }
+  
+  *length= stat_info.st_size;
+  
+  if (!(buffer= (uchar*)my_malloc(PSI_NOT_INSTRUMENTED, *length, MYF(MY_WME))))
+  {
+    fprintf(stderr, "Error: Cannot allocate memory for FRM file\n");
+    return NULL;
+  }
+  
+  if ((file= mysql_file_open(key_file_frm, filename, O_RDONLY | O_SHARE, MYF(0))) < 0)
+  {
+    fprintf(stderr, "Error: Cannot open file '%s': %s\n", 
+            filename, strerror(errno));
+    my_free(buffer);
+    return NULL;
+  }
+  
+  if (mysql_file_read(file, buffer, *length, MYF(MY_NABP)))
+  {
+    fprintf(stderr, "Error: Cannot read file '%s': %s\n", 
+            filename, strerror(errno));
+    my_free(buffer);
+    mysql_file_close(file, MYF(0));
+    return NULL;
+  }
+  
+  mysql_file_close(file, MYF(0));
+  return buffer;
+}
+
+/**
+  Extract database and table name from FRM file path
+*/
+static bool extract_db_table_names(const char *frm_path, 
+                                   LEX_CSTRING *db_name, 
+                                   LEX_CSTRING *table_name)
+{
+  char *path_copy, *db_start, *table_start, *frm_ext;
+  
+  if (!((path_copy= my_strdup(PSI_NOT_INSTRUMENTED, frm_path, MYF(MY_WME)))))
+    return true;
+  
+  if (!((frm_ext= strstr(path_copy, ".frm"))))
+  {
+    my_free(path_copy);
+    return true;
+  }
+  *frm_ext= '\0';
+  
+  table_start= strrchr(path_copy, '/');
+  if (!table_start)
+    table_start= strrchr(path_copy, '\\');
+  
+  if (!table_start)
+  {
+    table_name->str= my_strdup(PSI_NOT_INSTRUMENTED, path_copy, MYF(MY_WME));
+    table_name->length= strlen(table_name->str);
+    db_name->str= my_strdup(PSI_NOT_INSTRUMENTED, "test", MYF(MY_WME));
+    db_name->length= 4;
+    my_free(path_copy);
+    return false;
+  }
+  
+  *table_start= '\0';
+  table_start++;
+  
+  db_start= strrchr(path_copy, '/');
+  if (!db_start)
+    db_start= strrchr(path_copy, '\\');
+  
+  if (!db_start)
+    db_start= path_copy;
+  else
+    db_start++;
+  
+  table_name->str= my_strdup(PSI_NOT_INSTRUMENTED, table_start, MYF(MY_WME));
+  table_name->length= strlen(table_name->str);
+  db_name->str= my_strdup(PSI_NOT_INSTRUMENTED, db_start, MYF(MY_WME));
+  db_name->length= strlen(db_name->str);
+  
+  my_free(path_copy);
+  return false;
+}
+
+/**
+  Parse FRM file and create TABLE_SHARE and TABLE structures
+*/
+static bool parse_frm_file(THD *fake_thd, const char *frm_path)
+{
+  printf("DEBUG: Entering parse_frm_file\n");
+  fflush(stdout);
+  
+  TABLE_LIST table_list;
+  uchar *frm_data= NULL;
+  size_t frm_length= 0;
+  TABLE_SHARE *share= NULL;
+  TABLE *table= NULL;
+  LEX_CSTRING db_name, table_name;
+  bool error= true;
+  
+  memset(&table_list, 0, sizeof(table_list));
+  printf("DEBUG: table_list initialized\n");
+  fflush(stdout);
+
+  printf("DEBUG: About to read FRM file: %s\n", frm_path);
+  fflush(stdout);
+  
+  frm_data= read_frm_file(frm_path, &frm_length);
+  if (!frm_data)
+    goto cleanup;
```

Comment:
We have different opinions about gotos acros our communities, and mine is -- they are completely unnecessary, they make the code harder to read, and they make it impossible to add a variable initialization in-between the function (after a first goto).

Try to rewrite cleanups with the use of RAII. We have an ATEXIT convenience macro for that.
THD uses constructor and destructor, so use  unique_ptr for a THD pointer.